File: app/src/main/java/me/wcy/music/activity/MusicActivity.java
package me.wcy.music.activity;

import android.Manifest;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.graphics.Bitmap;
import android.media.AudioManager;
import android.os.Bundle;
import android.os.IBinder;
import android.support.design.widget.NavigationView;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.view.GravityCompat;
import android.support.v4.view.ViewPager;
import android.support.v4.widget.DrawerLayout;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import me.wcy.music.BuildConfig;
import me.wcy.music.R;
import me.wcy.music.adapter.FragmentAdapter;
import me.wcy.music.constants.Extras;
import me.wcy.music.executor.NaviMenuExecutor;
import me.wcy.music.executor.WeatherExecutor;
import me.wcy.music.fragment.LocalMusicFragment;
import me.wcy.music.fragment.PlayFragment;
import me.wcy.music.fragment.SongListFragment;
import me.wcy.music.model.Music;
import me.wcy.music.receiver.RemoteControlReceiver;
import me.wcy.music.service.OnPlayerEventListener;
import me.wcy.music.service.PlayService;
import me.wcy.music.utils.CoverLoader;
import me.wcy.music.utils.SystemUtils;
import me.wcy.music.utils.ToastUtils;
import me.wcy.music.utils.UpdateUtils;
import me.wcy.music.utils.binding.Bind;
import me.wcy.music.utils.permission.PermissionReq;
import me.wcy.music.utils.permission.PermissionResult;

public class MusicActivity extends BaseActivity implements View.OnClickListener, OnPlayerEventListener,
        NavigationView.OnNavigationItemSelectedListener, ViewPager.OnPageChangeListener {
    @Bind(R.id.drawer_layout)
    private DrawerLayout drawerLayout;
    @Bind(R.id.navigation_view)
    private NavigationView navigationView;
    @Bind(R.id.iv_menu)
    private ImageView ivMenu;
    @Bind(R.id.iv_search)
    private ImageView ivSearch;
    @Bind(R.id.tv_local_music)
    private TextView tvLocalMusic;
    @Bind(R.id.tv_online_music)
    private TextView tvOnlineMusic;
    @Bind(R.id.viewpager)
    private ViewPager mViewPager;
    @Bind(R.id.fl_play_bar)
    private FrameLayout flPlayBar;
    @Bind(R.id.iv_play_bar_cover)
    private ImageView ivPlayBarCover;
    @Bind(R.id.tv_play_bar_title)
    private TextView tvPlayBarTitle;
    @Bind(R.id.tv_play_bar_artist)
    private TextView tvPlayBarArtist;
    @Bind(R.id.iv_play_bar_play)
    private ImageView ivPlayBarPlay;
    @Bind(R.id.iv_play_bar_next)
    private ImageView ivPlayBarNext;
    @Bind(R.id.pb_play_bar)
    private ProgressBar mProgressBar;
    private View vNavigationHeader;
    private LocalMusicFragment mLocalMusicFragment;
    private SongListFragment mSongListFragment;
    private PlayFragment mPlayFragment;
    private PlayService mPlayService;
    private PlayServiceConnection mPlayServiceConnection;
    private AudioManager mAudioManager;
    private ComponentName mRemoteReceiver;
    private boolean isPlayFragmentShow = false;
    private MenuItem timerItem;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_music);

        bindService();
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        parseIntent(intent);
    }

    private void bindService() {
        Intent intent = new Intent();
        intent.setClass(this, PlayService.class);
        mPlayServiceConnection = new PlayServiceConnection();
        bindService(intent, mPlayServiceConnection, Context.BIND_AUTO_CREATE);
    }

    private class PlayServiceConnection implements ServiceConnection {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            mPlayService = ((PlayService.PlayBinder) service).getService();
            mPlayService.setOnPlayEventListener(MusicActivity.this);
            init();
            parseIntent(getIntent());
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
        }
    }

    private void init() {
        setupView();
        updateWeather();
        registerReceiver();
        onChange(mPlayService.getPlayingMusic());
        UpdateUtils.checkUpdate(this);
    }

    @Override
    protected void setListener() {
        ivMenu.setOnClickListener(this);
        ivSearch.setOnClickListener(this);
        tvLocalMusic.setOnClickListener(this);
        tvOnlineMusic.setOnClickListener(this);
        mViewPager.setOnPageChangeListener(this);
        flPlayBar.setOnClickListener(this);
        ivPlayBarPlay.setOnClickListener(this);
        ivPlayBarNext.setOnClickListener(this);
        navigationView.setNavigationItemSelectedListener(this);
    }

    private void setupView() {
        // add navigation header
        vNavigationHeader = LayoutInflater.from(this).inflate(R.layout.navigation_header, navigationView, false);
        navigationView.addHeaderView(vNavigationHeader);

        // setup view pager
        mLocalMusicFragment = new LocalMusicFragment();
        mSongListFragment = new SongListFragment();
        FragmentAdapter adapter = new FragmentAdapter(getSupportFragmentManager());
        adapter.addFragment(mLocalMusicFragment);
        adapter.addFragment(mSongListFragment);
        mViewPager.setAdapter(adapter);
        tvLocalMusic.setSelected(true);
    }

    private void updateWeather() {
        PermissionReq.with(this)
                .permissions(Manifest.permission.ACCESS_FINE_LOCATION,
                        Manifest.permission.ACCESS_COARSE_LOCATION)
                .result(new PermissionResult() {
                    @Override
                    public void onGranted() {
                        new WeatherExecutor(mPlayService, vNavigationHeader).execute();
                    }

                    @Override
                    public void onDenied() {
                        ToastUtils.show(getString(R.string.no_permission, "定位", "更新天气"));
                    }
                })
                .request();
    }

    private void registerReceiver() {
        mAudioManager = (AudioManager) getSystemService(AUDIO_SERVICE);
        mRemoteReceiver = new ComponentName(getPackageName(), RemoteControlReceiver.class.getName());
        mAudioManager.registerMediaButtonEventReceiver(mRemoteReceiver);
    }

    private void parseIntent(Intent intent) {
        if (intent.hasExtra(Extras.FROM_NOTIFICATION)) {
            showPlayingFragment();
        }
    }

    /**
     * 更新播放进度
     */
    @Override
    public void onPublish(int progress) {
        mProgressBar.setProgress(progress);
        if (mPlayFragment != null && mPlayFragment.isInitialized()) {
            mPlayFragment.onPublish(progress);
        }
    }

    @Override
    public void onChange(Music music) {
        onPlay(music);
        if (mPlayFragment != null && mPlayFragment.isInitialized()) {
            mPlayFragment.onChange(music);
        }
    }

    @Override
    public void onPlayerPause() {
        ivPlayBarPlay.setSelected(false);
        if (mPlayFragment != null && mPlayFragment.isInitialized()) {
            mPlayFragment.onPlayerPause();
        }
    }

    @Override
    public void onPlayerResume() {
        ivPlayBarPlay.setSelected(true);
        if (mPlayFragment != null && mPlayFragment.isInitialized()) {
            mPlayFragment.onPlayerResume();
        }
    }

    @Override
    public void onTimer(long remain) {
        if (timerItem == null) {
            timerItem = navigationView.getMenu().findItem(R.id.action_timer);
        }
        String title = getString(R.string.menu_timer);
        timerItem.setTitle(remain == 0 ? title : SystemUtils.formatTime(title + "(mm:ss)", remain));
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.iv_menu:
                drawerLayout.openDrawer(GravityCompat.START);
                break;
            case R.id.iv_search:
                startActivity(new Intent(this, SearchMusicActivity.class));
                break;
            case R.id.tv_local_music:
                mViewPager.setCurrentItem(0);
                break;
            case R.id.tv_online_music:
                mViewPager.setCurrentItem(1);
                break;
            case R.id.fl_play_bar:
                showPlayingFragment();
                break;
            case R.id.iv_play_bar_play:
                play();
                break;
            case R.id.iv_play_bar_next:
                next();
                break;
        }
    }

    @Override
    public boolean onNavigationItemSelected(final MenuItem item) {
        drawerLayout.closeDrawers();
        mHandler.postDelayed(new Runnable() {
            @Override
            public void run() {
                item.setChecked(false);
            }
        }, 500);
        return NaviMenuExecutor.onNavigationItemSelected(item, this);
    }

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
    }

    @Override
    public void onPageSelected(int position) {
        if (position == 0) {
            tvLocalMusic.setSelected(true);
            tvOnlineMusic.setSelected(false);
        } else {
            tvLocalMusic.setSelected(false);
            tvOnlineMusic.setSelected(true);
        }
    }

    @Override
    public void onPageScrollStateChanged(int state) {
    }

    public void onPlay(Music music) {
        if (music == null) {
            return;
        }
        Bitmap cover;
        if (music.getCover() == null) {
            cover = CoverLoader.getInstance().loadThumbnail(music.getCoverUri());
        } else {
            cover = music.getCover();
        }
        ivPlayBarCover.setImageBitmap(cover);
        tvPlayBarTitle.setText(music.getTitle());
        tvPlayBarArtist.setText(music.getArtist());
        if (getPlayService().isPlaying()) {
            ivPlayBarPlay.setSelected(true);
        } else {
            ivPlayBarPlay.setSelected(false);
        }
        mProgressBar.setMax((int) music.getDuration());
        mProgressBar.setProgress(0);

        if (mLocalMusicFragment != null && mLocalMusicFragment.isInitialized()) {
            mLocalMusicFragment.onItemPlay();
        }
    }

    private void play() {
        getPlayService().playPause();
    }

    private void next() {
        getPlayService().next();
    }

    public PlayService getPlayService() {
        return mPlayService;
    }

    private void showPlayingFragment() {
        FragmentTransaction ft = getSupportFragmentManager().beginTransaction();
        ft.setCustomAnimations(R.anim.fragment_slide_up, 0);
        if (mPlayFragment == null) {
            mPlayFragment = new PlayFragment();
            ft.replace(android.R.id.content, mPlayFragment);
        } else {
            ft.show(mPlayFragment);
        }
        ft.commit();
        isPlayFragmentShow = true;
    }

    private void hidePlayingFragment() {
        FragmentTransaction ft = getSupportFragmentManager().beginTransaction();
        ft.setCustomAnimations(0, R.anim.fragment_slide_down);
        ft.hide(mPlayFragment);
        ft.commit();
        isPlayFragmentShow = false;
    }

    @Override
    public void onBackPressed() {
        if (mPlayFragment != null && isPlayFragmentShow) {
            hidePlayingFragment();
            return;
        }
        if (drawerLayout.isDrawerOpen(GravityCompat.START)) {
            drawerLayout.closeDrawers();
            return;
        }

        if (BuildConfig.DEBUG) {
            super.onBackPressed();
        } else {
            moveTaskToBack(false);
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        // 切换夜间模式不保存状态
    }

    @Override
    protected void onDestroy() {
        if (mPlayServiceConnection != null) {
            unbindService(mPlayServiceConnection);
        }
        if (mRemoteReceiver != null) {
            mAudioManager.unregisterMediaButtonEventReceiver(mRemoteReceiver);
        }
        super.onDestroy();
    }
}

File: app/src/main/java/me/wcy/music/activity/MusicActivity.kt
package me.wcy.music.activity

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.TextView
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.viewpager.widget.ViewPager
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import com.google.android.material.navigation.NavigationView
import me.wcy.common.permission.Permissioner
import me.wcy.music.R
import me.wcy.music.adapter.FragmentAdapter
import me.wcy.music.constants.Extras
import me.wcy.music.constants.Keys
import me.wcy.music.executor.ControlPanel
import me.wcy.music.executor.NaviMenuExecutor
import me.wcy.music.fragment.LocalMusicFragment
import me.wcy.music.fragment.PlayFragment
import me.wcy.music.fragment.SheetListFragment
import me.wcy.music.service.AudioPlayer
import me.wcy.music.service.PlayService
import me.wcy.music.service.QuitTimer
import me.wcy.music.service.QuitTimer.OnTimerListener
import me.wcy.music.utils.SystemUtils
import me.wcy.music.utils.binding.Bind

class MusicActivity : BaseActivity(), View.OnClickListener, OnTimerListener,
    NavigationView.OnNavigationItemSelectedListener, OnPageChangeListener {
    @Bind(R.id.drawer_layout)
    private val drawerLayout: DrawerLayout? = null

    @Bind(R.id.navigation_view)
    private val navigationView: NavigationView? = null

    @Bind(R.id.iv_menu)
    private val ivMenu: ImageView? = null

    @Bind(R.id.iv_search)
    private val ivSearch: ImageView? = null

    @Bind(R.id.tv_local_music)
    private val tvLocalMusic: TextView? = null

    @Bind(R.id.tv_online_music)
    private val tvOnlineMusic: TextView? = null

    @Bind(R.id.viewpager)
    private val mViewPager: ViewPager? = null

    @Bind(R.id.fl_play_bar)
    private val flPlayBar: FrameLayout? = null
    private var vNavigationHeader: View? = null
    private var mLocalMusicFragment: LocalMusicFragment? = null
    private var mSheetListFragment: SheetListFragment? = null
    private var mPlayFragment: PlayFragment? = null
    private val controlPanel: ControlPanel by lazy {
        ControlPanel(flPlayBar)
    }
    private var naviMenuExecutor: NaviMenuExecutor? = null
    private var timerItem: MenuItem? = null
    private var isPlayFragmentShow = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_music)

        val intent = Intent(this, PlayService::class.java)
        startService(intent)

        Permissioner.requestNotificationPermission(this, null)

        setupView()
        naviMenuExecutor = NaviMenuExecutor(this)
        AudioPlayer.get().addOnPlayEventListener(controlPanel)
        QuitTimer.get().setOnTimerListener(this)
        parseIntent()
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        setIntent(intent)
        parseIntent()
    }

    private fun setupView() {
        // add navigation header
        vNavigationHeader =
            LayoutInflater.from(this).inflate(R.layout.navigation_header, navigationView, false)
        navigationView!!.addHeaderView(vNavigationHeader!!)

        // setup view pager
        mLocalMusicFragment = LocalMusicFragment()
        mSheetListFragment = SheetListFragment()
        val adapter = FragmentAdapter(supportFragmentManager)
        adapter.addFragment(mLocalMusicFragment!!)
        adapter.addFragment(mSheetListFragment!!)
        mViewPager!!.adapter = adapter
        tvLocalMusic!!.isSelected = true
        ivMenu!!.setOnClickListener(this)
        ivSearch!!.setOnClickListener(this)
        tvLocalMusic.setOnClickListener(this)
        tvOnlineMusic!!.setOnClickListener(this)
        flPlayBar!!.setOnClickListener(this)
        mViewPager.addOnPageChangeListener(this)
        navigationView.setNavigationItemSelectedListener(this)
    }

    private fun parseIntent() {
        val intent = intent
        if (intent.hasExtra(Extras.EXTRA_NOTIFICATION)) {
            showPlayingFragment()
            setIntent(Intent())
        }
    }

    override fun onTimer(remain: Long) {
        if (timerItem == null) {
            timerItem = navigationView!!.menu.findItem(R.id.action_timer)
        }
        val title = getString(R.string.menu_timer)
        timerItem!!.title = if (remain == 0L) title else SystemUtils.formatTime(
            "$title(mm:ss)",
            remain
        )
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.iv_menu -> drawerLayout!!.openDrawer(GravityCompat.START)
            R.id.iv_search -> {}
            R.id.tv_local_music -> mViewPager!!.currentItem = 0
            R.id.tv_online_music -> mViewPager!!.currentItem = 1
            R.id.fl_play_bar -> showPlayingFragment()
        }
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        drawerLayout!!.closeDrawers()
        handler!!.postDelayed({ item.isChecked = false }, 500)
        return naviMenuExecutor!!.onNavigationItemSelected(item)
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}
    override fun onPageSelected(position: Int) {
        if (position == 0) {
            tvLocalMusic!!.isSelected = true
            tvOnlineMusic!!.isSelected = false
        } else {
            tvLocalMusic!!.isSelected = false
            tvOnlineMusic!!.isSelected = true
        }
    }

    override fun onPageScrollStateChanged(state: Int) {}
    private fun showPlayingFragment() {
        if (isPlayFragmentShow) {
            return
        }
        val ft = supportFragmentManager.beginTransaction()
        ft.setCustomAnimations(R.anim.fragment_slide_up, 0)
        if (mPlayFragment == null) {
            mPlayFragment = PlayFragment()
            ft.replace(android.R.id.content, mPlayFragment!!)
        } else {
            ft.show(mPlayFragment!!)
        }
        ft.commitAllowingStateLoss()
        isPlayFragmentShow = true
    }

    private fun hidePlayingFragment() {
        val ft = supportFragmentManager.beginTransaction()
        ft.setCustomAnimations(0, R.anim.fragment_slide_down)
        ft.hide(mPlayFragment!!)
        ft.commitAllowingStateLoss()
        isPlayFragmentShow = false
    }

    override fun onBackPressed() {
        if (mPlayFragment != null && isPlayFragmentShow) {
            hidePlayingFragment()
            return
        }
        if (drawerLayout!!.isDrawerOpen(GravityCompat.START)) {
            drawerLayout.closeDrawers()
            return
        }
        super.onBackPressed()
    }

    @SuppressLint("MissingSuperCall")
    override fun onSaveInstanceState(outState: Bundle) {
        outState.putInt(Keys.VIEW_PAGER_INDEX, mViewPager!!.currentItem)
        mLocalMusicFragment?.onSaveInstanceState(outState)
        mSheetListFragment?.onSaveInstanceState(outState)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        mViewPager?.post {
            mViewPager.setCurrentItem(
                savedInstanceState.getInt(Keys.VIEW_PAGER_INDEX),
                false
            )
            mLocalMusicFragment?.onRestoreInstanceState(savedInstanceState)
            mSheetListFragment?.onRestoreInstanceState(savedInstanceState)
        }
    }

    override fun onDestroy() {
        AudioPlayer.get().removeOnPlayEventListener(controlPanel)
        QuitTimer.get().setOnTimerListener(null)
        super.onDestroy()
    }
}